overlord/ifacestate/ifacestate.go

```diff
@@ -39,13 +40,78 @@ var noConflictOnConnectTasks = func(task *state.Task) bool {
 	return task.Kind() != "connect" && task.Kind() != "disconnect"
 }
 
+var connectRetryTimeout = time.Second * 5
+
+func checkConnectConflicts(st *state.State, change *state.Change, plugSnap, slotSnap string, autoConnectTask *state.Task) error {
+	for _, chg := range st.Changes() {
+		if chg.Kind() == "transition-ubuntu-core" {
+			return fmt.Errorf("ubuntu-core to core transition in progress, no other changes allowed until this is done")
+		}
+	}
+
+	var installedSnap string
+	if autoConnectTask != nil {
+		snapsup, err := snapstate.TaskSnapSetup(autoConnectTask)
+		if err != nil {
+			return fmt.Errorf("internal error: cannot obtain snap setup from task: %s", autoConnectTask.Summary())
+		}
+		installedSnap = snapsup.Name()
+	}
+
+	for _, task := range st.Tasks() {
+		if task.Status().Ready() || autoConnectTask == task {
+			continue
+		}
+
+		k := task.Kind()
+		if k == "connect" || k == "disconnect" {
+			continue
+		}
+
+		snapsup, err := snapstate.TaskSnapSetup(task)
+		// e.g. hook tasks don't have task snap setup
+		if err != nil {
+			continue
+		}
+
+		snapName := snapsup.Name()
+
+		if autoConnectTask != nil && installedSnap == snapName {
+			continue
+		}
+
+		// different snaps - no conflict
+		if snapName != plugSnap && snapName != slotSnap {
+			continue
+		}
+
+		if k == "unlink-snap" || k == "link-snap" || k == "setup-profiles" {
+			if autoConnectTask != nil {
+				// if snap is getting removed, we will retry but the snap will be gone and auto-connect becomes no-op
+				// if snap is getting installed/refreshed - temporary conflict, retry later
+				return &state.Retry{After: connectRetryTimeout}
```

Comment:
it would be good to have more of an integration test for this (maybe in overlord/managers_test.go) where we install at the same time (same change) two snaps that will also auto-connect

Reply:
Added.